content/blog/posts/o3de-motion-matching-2022-04.md

```diff
@@ -0,0 +1,179 @@
+
+The feature schema defines what the cost function looks like. Every feature added to the feature schema adds to the cost. The bigger the discrepancy between the current velocity and the one from the frame candidate, the higher the penalty to the cost and the less likely the candidate is a good one to take.
+
+This makes motion matching an optimization problem where the frame with the minimum cost is the most preferred candidate to transition to.
+
+#### Searching next best matching frame
+
+The actual search happens in two phases, a broad phase to eliminate most of the candidates followed by a narrow phase to find the actual best candidate.
+
+##### 1. Broad-phase (KD-tree)
+
+A KD-tree is used to find the nearest neighbors (frames in the motion database) to the query vector (given input). The result is a set of pre-selected frames for the next best matching frame that is passed on to the narrow-phase. By adjusting the maximum tree depth or the minimum number of frames for the leaf nodes, the resulting number of frames can be adjusted. The bigger the set of frames the broad-phase returns, the more candidates the narrow-phase can choose from, the better the visual quality of the animation but the slower the algorithm.
+
+##### 2. Narrow-phase
+
+Inside the narrow-phase, we iterate through the returned set of frames from the KD-tree, and evaluate and compare their cost against each other. The frame with the minimal cost is the best match that we transition to.
+
+Pseudo-code:
+
+```
+minCost = MAX;
+for_all (nearest frames found in the broad-phase)
+{
+	frameCost = 0.0
+	for_all (features)
+	{
+		frameCost += CalculateCost(feature);
+	}
+
+	if (frameCost < minCost)
+	{
+		// We found a better next matching frame
+		minCost = frameCost;
+		newBestMatchingFrame = currentFrame;
+	}
+}
+
+StartTransition(newBestMatchingFrame);
+```
+
+## Character trajectories
+
+Let's talk about trajectories. A trajectory is a useful way to encode where the character came from and its intention of where and how it wants to move in the future. We store a number of samples containing character positions as well as facing directions for a given time window. For example, 4 samples into the past across 0.7 seconds and 6 samples for the desired future trajectory across 1.2 seconds. The Euclidean distances between the samples tell us if the character is accelerating or coming to a stop. The sample locations indicate if it is moving forward or doing a left turn, and the facing direction gives us more information about if the character is strafing or moving backwards. Recording the past trajectory is easy as we know the current character transform, but how do we decide the future trajectory to generate, and why do we even need that?
```

Comment:
```suggestion
Let's talk about trajectories. A trajectory is a useful way to encode where the character came from and its intention of where and how it wants to move in the future. We store a number of samples containing character positions, as well as facing directions, for a given time window. For example, 4 samples into the past across 0.7 seconds and 6 samples for the desired future trajectory across 1.2 seconds. The Euclidean distances between the samples tell us if the character is accelerating or coming to a stop. The sample locations indicate if the character is moving forward or doing a left turn, and the facing direction gives us more information about if the character is strafing or moving backwards. Recording the past trajectory is easy as we know the current character transform, but how do we decide the future trajectory to generate, and why do we even need that?
```

Reply:
minor: Its odd to end in a questions esp when the content is immediately followed by an image.